File: 0042-trapping-rain-water/solution.cs
public class Solution {
    public int Trap(int[] height) {
        int trapped = 0;
        int l = 0;
        int r = height.Length -1;
        int lmax = height[l];
        int rmax = height[r];
        while(l<r){
            if (lmax < rmax){
                l +=1;
                lmax = Math.Max(lmax, height[l]);
                trapped += lmax - height[l];
            }
            else{
                r -=1;
                rmax = Math.Max(rmax, height[r]);
                trapped += rmax - height[r];
            }
        }
        return trapped;
    }
}

File: 0042-trapping-rain-water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        lmax = [0]*(len(height))
        rmax = [0]*(len(height))
        ans = 0
        lmax[0] = height[0]
        rmax[len(height)-1] = height[len(height)-1]
        for i in range(1, len(height)):
            lmax[i] = max(lmax[i-1], height[i])

        for i in range(len(height)-2, -1, -1):
            rmax[i] = max(rmax[i+1], height[i])

        for i in range(len(height)):
            ans+= min(lmax[i], rmax[i])- height[i]
        return ans
